app/shutdownhandler.go

```diff
@@ -0,0 +1,188 @@
+package app
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/arquivei/foundationkit/errors"
+	"github.com/rs/zerolog/log"
+)
+
+// ErrorPolicy specifies what should be done when a handler fails
+type ErrorPolicy int
+
+// ShutdownPriority is used to guide the execution of the shutdown handlers
+// during a graceful shutdown. The shutdown is performed from the higher to the lowest
+// priority
+type ShutdownPriority int
+
+const (
+	// ErrorPolicyWarn prints the error as a warning and continues to the next handler. This is the default.
+	ErrorPolicyWarn ErrorPolicy = iota
+	// ErrorPolicyAbort stops the shutdown process and returns an error
+	ErrorPolicyAbort
+	// ErrorPolicyFatal logs the error as Fatal, it means the application will close immediately
+	ErrorPolicyFatal
+	// ErrorPolicyPanic panics if there is an error
+	ErrorPolicyPanic
+)
+
+// ShutdownFunc is a shutdown function that will be executed when the app is shutting down.
+type ShutdownFunc func(context.Context) error
+
+type shutdownHandler struct {
+	Name     string
+	Timeout  time.Duration
+	Policy   ErrorPolicy
+	Priority ShutdownPriority
+
+	Shutdown ShutdownFunc
+	executed bool
+	err      error
+	mu       sync.Mutex
+	index    int
+}
+
+// Used to set priority automatically by insertion order
+// This is to maintain consistency with the previous queue behavior
+var nextErrorPriority = 0
+var errorPriorityMode = errorPriorityModeUnset
+
+const (
+	errorPriorityModeUnset = iota
+	errorPriorityModeInsertion
+	errorPriorityModeOption
+)
+
+// NewShutdownHandler returns a new
+func NewShutdownHandler(name string, fn ShutdownFunc, options ...interface{}) *shutdownHandler {
+	if name == "" {
+		panic("shutdown handler must have a name")
+	}
+	sh := &shutdownHandler{
+		Name:     name,
+		Shutdown: fn,
+	}
+	for _, o := range options {
+		switch opt := o.(type) {
+		case time.Duration:
+			sh.Timeout = opt
+		case ErrorPolicy:
+			sh.Policy = opt
+		case ShutdownPriority:
+			if errorPriorityMode == errorPriorityModeUnset {
```

Comment:
Nit. This is not wrong, but I disagree with using `javascript parameter pattern` (arbitrary number of any-typed parameters) together with a "configure only once" behavior that might or might not be triggered. I'd prefer a `Setup` function that must be called one time only.

Reply:
I don't know if it is related, but I'd rather use a Config structure than this `options ...interface{}` argument, since it would be type safe and self documented

Reply:
I see no problem with a config struct, but I wouldn't change this now, but in a future release.

Reply:
> Nit. This is not wrong, but I disagree with using `javascript parameter pattern` (arbitrary number of any-typed parameters) together with a "configure only once" behavior that might or might not be triggered. I'd prefer a `Setup` function that must be called one time only.

I was trying to keep the old code running without change. If we require a setup process, it would be better only have the new hehavior and refactor all clients using this code.